### ISSUE TITLE
The rhythm step freezes the phone while it runs

### ISSUE THREAD

**Issue by popelenkow** [musetric/musetric#933]
### Problem

While the rhythm step runs, the phone stops drawing for long stretches: the shade, the status bar and other apps wait for the GPU. #894 describes the same problem for the RoFormer core and #903 for Whisper; #896 and #904 fixed those by pacing the submissions on a device we own (#897). The rhythm step is not paced, and it submits its feature passes as one dispatch each over the whole track, so a job from any other client of the GPU waits for whatever the step has queued.

### What is measured

- **Step time:** one `analyze` call of the product runtime over the whole track, which is what the host feeds as a single unit.
- **Foreign job wait:** a one-dispatch compute job submitted every 100 ms from a worker, on a GPU device of its own, while the step runs. The wait is measured from `queue.submit` to `onSubmittedWorkDone`. This is what any other client of the GPU waits for its turn, the compositor included. Probes still queued when the step ends are waited for and counted, since they are the longest waits.
- **Model load:** `createBeatThisGpuRuntime` with the model file already in the browser cache.
- **Output checksum:** the sum of the beat logits, so a change that moves a number shows up.

The bench calls the product runtime (`createBeatThisGpuRuntime`, `analyze`) from the commit being measured, with the graph the Rust host declares and the model revision the app pins (`a076df6f20345e133a73b6d2068b68b60e48fafb`).

The input is Rxbyn, "Bad Side" (https://www.jamendo.com/track/1556580/bad-side, CC BY 3.0, from the JamendoLyrics dataset), the whole track, mono 22.05 kHz - 216 s, which is 10 830 frames and 22 model windows.

### How the measurement is run

One **load** is a fresh page and a fresh runtime: load the model, run the step once, release. One **measurement** is three loads per device with 120 s idle between loads, and the device temperature is read before and after each load. The phone's screen is on and unlocked for the whole load; the desktop runs with `--force_high_performance_gpu`.

### State at `58328cbd`

Measured on 2026-09-21, three loads per device, the 9RT at 78 °C and the S24 at 45 °C throughout.

| | RTX 3060 Laptop | Adreno 750 | Adreno 660 |
| --- | --- | --- | --- |
| Step time | 1220 / 1174 / 1183 ms | 4937 / 4642 / 4666 ms | 11 301 / 11 091 / 11 133 ms |
| Model load | 1946 ms | 5958 ms | 7665 ms |
| Foreign job wait, p50 | 10 ms | 109 ms | 337 ms |
| Foreign job wait, p95 | 16 ms | 216 ms | 596 ms |
| Foreign job wait, worst | 18 ms | 240 ms | 620 ms |

On the Adreno 660 a foreign job waits a third of a second in the median for the eleven seconds the step runs, and two thirds of the probes wait longer than 250 ms. The Adreno 750 stays under a quarter of a second, and the desktop is inside one frame.

An idle page on the same phones, with the probe running and nothing else submitted, waits 11 ms at the median, 13 ms at p95 and 15 ms at worst. That is the floor of the instrument.

Where it comes from: the step submits the frame pass, the transform, the mel pass and the window gather as one dispatch each over the whole track, and then runs the model window by window without a limit on what the queue holds. Timed one submission at a time on the Adreno 660, the mel pass alone is 105 ms and the frame pass 64 ms, while the transform is 5 to 8 ms per batch of frames.

### The chord step, for comparison

The same measurement on the chord step, which is the other unpaced step: the Adreno 660 takes 839 to 1002 ms for the whole track and a foreign job waits 60 ms in the median, with the worst of three loads at 210 ms; the Adreno 750 takes 386 to 562 ms and waits 46 ms in the median, worst 67 ms. The step is short enough that a load collects only six to nine probes, so the tail is thin evidence. It is not in the same class as the rhythm step and is not part of this issue.

### Scripts

The bench page is built from the monorepo at the commit being measured, so it runs the product runtime. `canaryWorker.ts` and the results server are the ones published in #894 and #903.

<details>
<summary>bench page</summary>

```ts
import { verifyCqtPlanArtifact } from '@musetric/cqt';
import * as ort from 'onnxruntime-web/webgpu';
import { createChordNetGpuRuntime } from '../../../../../musetric/packages/ai/src/runtime/chords/chordNetGpuRuntime.js';
import { serveOrtWasmFromBundle } from '../../../../../musetric/packages/ai/src/runtime/ortWasm.js';
import { getMusetricWebGpuDevice } from '../../../../../musetric/packages/ai/src/runtime/webgpuDevice.js';
import { createBeatThisGpuRuntime } from '../../../../../musetric/packages/ai/src/runtime/rhythm/beatThisGpuRuntime.js';

const params = new URLSearchParams(location.search);
const tag = params.get('tag') ?? 'steps';
const step = params.get('step') ?? 'rhythm';
const clip = params.get('clip') ?? 'clip22.f32';
const seconds = Number(params.get('seconds') ?? '60');
const runs = Number(params.get('runs') ?? '1');
// Timing every submission costs a fence per submission, so it is opt-in.
const timing = params.get('timing') === '1';
const profile = params.get('profile') === '1';
const rhythmModel = params.get('model') ?? 'beat_this.onnx';

const sampleRate = 22050;
const models = 'models/steps';

// The graphs the Rust host hands the runtimes (analysis/models.rs).
const chordNetGraph = {
  inputName: 'features',
  outputName: 'logits',
  frameDuration: 2048 / 22050,
  sequenceLength: 108,
  inputBins: 144,
  chordCount: 170,
  windowsPerRun: 16,
};
const beatThisGraph = {
  inputName: 'spect',
  beatOutputName: 'beat',
  downbeatOutputName: 'downbeat',
  nFft: 1024,
  hopLength: 441,
  fps: 50,
  melBins: 128,
  logMultiplier: 1000,
  chunkSize: 513,
  borderSize: 6,
};

const lines: string[] = [];
const say = (text: string): void => {
  lines.push(text);
  (document.getElementById('log') as HTMLElement).textContent = lines.join('\n');
};
const post = async (payload: Record<string, unknown>): Promise<unknown> =>
  fetch('/result', { method: 'POST', body: JSON.stringify({ tag, step, seconds, device: navigator.userAgent, ...payload }) });

type KernelRecord = { kernelType: string; programName: string; startTime: bigint; endTime: bigint };
const kernels: KernelRecord[] = [];

const run = async (): Promise<void> => {
  serveOrtWasmFromBundle();
  if (profile) {
    // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
    (ort.env.webgpu as unknown as { profiling: unknown }).profiling = {
      mode: 'default',
      ondata: (data: KernelRecord) => kernels.push(data),
    };
  }
  const whole = new Float32Array(await (await fetch('/' + clip)).arrayBuffer());
  const audio = whole.subarray(0, Math.min(whole.length, Math.round(seconds * sampleRate)));

  // The foreign job runs in a worker on a GPU device of its own, like any other
  // client of this GPU would.
  const worker = new Worker(new URL('./canaryWorker.ts', import.meta.url), { type: 'module' });
  const canary = new Promise<{ waits: number[]; started: number }>((resolve, reject) => {
    worker.onmessage = (event: MessageEvent<{ kind: string; waits: number[]; started: number; message: string }>): void => {
      if (event.data.kind === 'report') resolve(event.data);
      if (event.data.kind === 'error') reject(new Error(event.data.message));
    };
  });

  // Every submission is timed from the outside: whatever the step submits, a
  // foreign client of this GPU waits that long for its own frame.
  const { device } = await getMusetricWebGpuDevice();
  const { queue } = device;
  const submit = queue.submit.bind(queue);
  const submittedWorkDone = queue.onSubmittedWorkDone.bind(queue);
  const submissions: { at: number; ms: number }[] = [];
  if (timing) {
    queue.submit = (buffers) => {
      const at = performance.now();
      submit(buffers);
      void submittedWorkDone().then(() => submissions.push({ at, ms: performance.now() - at }));
    };
  }

  if (step === 'idle') {
    worker.postMessage('start');
    await new Promise((resolve) => setTimeout(resolve, seconds * 1000));
    worker.postMessage('stop');
    worker.postMessage('report');
    const idle = await canary;
    await post({ stage: 'verdict', loadMs: 0, runMs: [0], checksums: [0], canaryWaitMs: idle.waits, canaryStarted: idle.started });
    say('idle done');
    return;
  }

  const loadAt = performance.now();
  const runtime =
    step === 'chords'
      ? await createChordNetGpuRuntime({
          graph: chordNetGraph,
          modelUrl: `/${models}/chordnet.onnx`,
          plan: await verifyCqtPlanArtifact(new Uint8Array(await (await fetch(`/${models}/cqt-plan.bin`)).arrayBuffer())),
        })
      : await createBeatThisGpuRuntime({
          graph: beatThisGraph,
          modelUrl: `/${models}/${rhythmModel}`,
          filterbank: new Float32Array(await (await fetch(`/${models}/mel-filterbank.bin`)).arrayBuffer()),
        });
  const loadMs = performance.now() - loadAt;
  say(`load ${(loadMs / 1000).toFixed(1)} s`);

  worker.postMessage('start');
  const runMs: number[] = [];
  const checksums: number[] = [];
  for (let index = 0; index < runs; index += 1) {
    const at = performance.now();
    const output = await runtime.analyze(audio);
    runMs.push(performance.now() - at);
    const values = 'beat' in output ? output.beat : output;
    let sum = 0;
    for (const value of values) sum += Number(value);
    checksums.push(sum);
    say(`run ${index}: ${runMs[index].toFixed(0)} ms`);
  }
  worker.postMessage('stop');
  worker.postMessage('report');
  const { waits, started } = await canary;
  await runtime.release();
  await submittedWorkDone();
  queue.submit = submit;
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const phases = (globalThis as unknown as { __phases?: Record<string, number[]> }).__phases ?? {};
  // eslint-disable-next-line @typescript-eslint/consistent-type-assertions
  const fences = (globalThis as unknown as { __fences?: number[]; __subs?: number[] });
  const byProgram: Record<string, { count: number; totalUs: number; maxUs: number }> = {};
  for (const kernel of kernels) {
    const us = Number(kernel.endTime - kernel.startTime) / 1000;
    const entry = (byProgram[`${kernel.kernelType}/${kernel.programName}`] ??= { count: 0, totalUs: 0, maxUs: 0 });
    entry.count += 1;
    entry.totalUs += us;
    entry.maxUs = Math.max(entry.maxUs, us);
  }
  await post({ stage: 'verdict', byProgram, fences: fences.__fences, subs: fences.__subs, phases, loadMs, runMs, checksums, canaryWaitMs: waits, canaryStarted: started, submissionMs: submissions.map((item) => Math.round(item.ms)) });
  say('done');
};

void run().catch(async (error: unknown) => {
  say(String(error));
  await post({ stage: 'fatal', error: String((error as Error)?.stack ?? error) });
});
```

</details>

<details>
<summary>one load</summary>

```js
// One load of the rhythm/chords freeze bench: a fresh page analyses the clip once
// while a worker measures the foreign job wait.
//   node steps-run.mjs <pc|adb serial> <tag> <rhythm|chords> [seconds]
import { execFileSync, spawn } from 'node:child_process';
import { appendFileSync, readFileSync } from 'node:fs';

const [target, tag, step = 'rhythm', seconds = '60', page = 'steps'] = process.argv.slice(2);
const wait = (ms) => Atomics.wait(new Int32Array(new SharedArrayBuffer(4)), 0, 0, ms);
const read = () =>
  readFileSync('freeze-results.jsonl', 'utf8').trim().split('\n').flatMap((line) => {
    try { return [JSON.parse(line)]; } catch { return []; }
  });
const adb = (args) => execFileSync('adb', ['-s', target, ...args], { encoding: 'utf8' });
const temperature = () => {
  if (target === 'pc') return undefined;
  const text = adb(['shell', 'dumpsys', 'thermalservice']);
  const values = [];
  for (const match of text.matchAll(/Temperature\{mValue=([\d.-]+), mType=(\d+), mName=([^,]+), mStatus=(\d+)\}/g)) {
    if (match[2] === '0' || match[2] === '3') values.push(Number(match[1]));
  }
  return values.length ? Math.max(...values) : undefined;
};
const quantile = (values, q) => {
  if (values.length === 0) return undefined;
  const sorted = [...values].sort((a, b) => a - b);
  return Math.round(sorted[Math.min(sorted.length - 1, Math.floor(q * sorted.length))]);
};

const url = `http://127.0.0.1:8099/${page}/index.html?tag=${tag}&step=${step}&seconds=${seconds}&timing=${process.env.TIMING ?? "0"}&profile=${process.env.PROFILE ?? "0"}&model=${process.env.MODEL ?? "beat_this.onnx"}`;
const before = read().length;
const temperatureBefore = temperature();
let chrome;
if (target === 'pc') {
  chrome = spawn('C:/Program Files/Google/Chrome/Application/chrome.exe', [`--user-data-dir=${process.env.TEMP}/steps-profile`, '--force_high_performance_gpu', '--no-first-run', '--no-default-browser-check', '--disable-background-timer-throttling', '--disable-backgrounding-occluded-windows', '--disable-renderer-backgrounding', url], { stdio: 'ignore' });
} else {
  adb(['shell', 'am', 'force-stop', 'com.android.chrome']);
  adb(['shell', 'input', 'keyevent', 'KEYCODE_WAKEUP']);
  adb(['shell', 'wm', 'dismiss-keyguard']);
  wait(2000);
  execFileSync('adb', ['-s', target, 'shell', 'am', 'start', '-a', 'android.intent.action.VIEW', '-d', `'${url}'`, 'com.android.chrome'], { stdio: 'ignore' });
}
const startedAt = Date.now();
let record;
while (Date.now() - startedAt < 1800000) {
  wait(5000);
  record = read().slice(before).find((row) => row.tag === tag && (row.stage === 'verdict' || row.stage === 'fatal'));
  if (record) break;
}
if (chrome) { try { execFileSync('taskkill', ['/F', '/T', '/PID', String(chrome.pid)], { stdio: 'ignore' }); } catch { /* gone */ } }
else { try { adb(['shell', 'am', 'force-stop', 'com.android.chrome']); } catch { /* wedged */ } }
const waits = record?.canaryWaitMs ?? [];
const load = {
  tag, target, step, page, seconds: Number(seconds), stage: record?.stage ?? 'timeout',
  temperatureBefore, temperatureAfter: temperature(),
  loadMs: record?.loadMs && Math.round(record.loadMs), runMs: record?.runMs?.map(Math.round),
  checksums: record?.checksums, probes: waits.length, started: record?.canaryStarted,
  waitP50: quantile(waits, 0.5), waitP95: quantile(waits, 0.95), waitMax: waits.length ? Math.round(Math.max(...waits)) : undefined,
  over250: waits.filter((value) => value > 250).length,
  submissions: record?.submissionMs?.length,
  subP50: quantile(record?.submissionMs ?? [], 0.5), subP95: quantile(record?.submissionMs ?? [], 0.95),
  subMax: record?.submissionMs?.length ? Math.max(...record.submissionMs) : undefined,
  subOver60: record?.submissionMs?.filter((value) => value > 60).length,
  error: record?.error?.slice(0, 300),
};
appendFileSync('steps-loads.jsonl', `${JSON.stringify({ ...load, canaryWaitMs: waits, submissionMs: record?.submissionMs })}\n`);
console.log(JSON.stringify(load));
```

</details>

<details>
<summary>one measurement</summary>

```sh
#!/usr/bin/env bash
# A-B-B-A-A-B over the two builds, with the device idle between loads.
#   ./steps-matrix.sh <pc|serial> <name> <rhythm|chords> <seconds> <idle seconds>
set -e
target=$1; name=$2; step=$3; seconds=$4; idle=${5:-120}
i=0
for page in steps-base steps steps steps-base steps-base steps; do
  i=$((i + 1))
  node steps-run.mjs "$target" "${name}-${i}-${page}" "$step" "$seconds" "$page" | tail -1
  sleep "$idle"
done
```

</details>

### Version

`58328cbd`

### Platform

Android

### Environment

- NVIDIA GeForce RTX 3060 Laptop, Chrome 153, `--force_high_performance_gpu`
- Galaxy S24 Ultra (SM-S928B, Adreno 750, Android 16), Chrome 153
- OnePlus 9RT (MT2111, Adreno 660, Android 14), Chrome 152

Both phones are served over `adb reverse`, onnxruntime-web 1.30.0.

**Comment by popelenkow** [comment]
### State at `2ac200f1`

The same measurement as above, three loads of each build per device, alternating, at a steady device temperature. "Before" is `58328cbd`, "after" is the merge.

| | RTX 3060 Laptop | Adreno 750 | Adreno 660 |
| --- | --- | --- | --- |
| Step time, before | 1220 / 1174 / 1183 ms | 4937 / 4642 / 4666 ms | 11 301 / 11 091 / 11 133 ms |
| Step time, after | 1419 / 1862 / 1622 ms | 5661 / 5757 / 5717 ms | 14 500 / 14 271 / 14 539 ms |
| Foreign job wait, p50 | 10 -> 5 ms | 109 -> 13 ms | 337 -> 14 ms |
| Foreign job wait, p95 | 16 -> 16 ms | 216 -> 37 ms | 596 -> 42 ms |
| Foreign job wait, worst | 18 -> 19 ms | 240 -> 97 ms | 620 -> 87 ms |

The step's output checksum is unchanged on every device and in every load.

What did the work: the pacer waits once the queue holds about 32 ms of submissions instead of after every one, the frame and the mel pass run in batches of frames from a uniform slot of their own, and the audio is uploaded in pieces.

What did not, and is not in the change: waiting after every submission holds the worst at 53 to 60 ms on the Adreno 660, but doubles the step on both phones and takes the desktop from 1.2 s to 4.8 s; splitting the three frontend convolutions of the graph into row bands at export, measured at 15 and 30 bands on device, stays within the run-to-run spread; halving the feature batches again does nothing.

What is left: the median and p95 are now at the floor of the instrument, which an idle page puts at 11 ms and 13 ms, but a load still carries one or two probes at 87 ms on the Adreno 660 and 97 ms on the Adreno 750. Those are not queue depth - they survive every policy measured - and the next thing that would move them is the flush threshold of the execution provider, which is not reachable from JavaScript (#893).
